File: stemqa/backend/naming.py
from __future__ import annotations


def _compact(value: str, strip_periods: bool = False, max_length: int | None = None) -> str:
    text = str(value or "")
    if strip_periods:
        text = text.replace(".", "")
    text = text.replace(" ", "")
    if max_length is not None:
        text = text[:max_length]
    return text


def build_name(
    catalog_id: str,
    composer: str,
    title: str,
    instrument: str,
    stem_type: str,
    version: int,
    sr: int,
    bit_depth: int,
) -> str:
    parts = [
        "TOMPLAY",
        _compact(catalog_id),
        _compact(composer),
        _compact(title, strip_periods=True, max_length=12),
        _compact(instrument),
        str(stem_type),
        f"v{str(version).zfill(2)}",
        f"{sr // 1000}k",
        f"{bit_depth}b",
    ]
    return "_".join(parts) + ".wav"

File: stemqa/frontend/src/lib/naming.js
function compact(value, { stripPeriods = false, maxLength } = {}) {
  let nextValue = String(value ?? '')
  if (stripPeriods) {
    nextValue = nextValue.replaceAll('.', '')
  }
  nextValue = nextValue.replaceAll(' ', '')
  if (typeof maxLength === 'number') {
    nextValue = nextValue.slice(0, maxLength)
  }
  return nextValue
}

export function buildName({
  catalogId,
  composer,
  title,
  instrument,
  stemType,
  version,
  sampleRate,
  bitDepth,
}) {
  const parts = [
    'TOMPLAY',
    compact(catalogId || 'UNKNOWN'),
    compact(composer || 'UNKNOWN'),
    compact(title || 'UNTITLED', { stripPeriods: true, maxLength: 12 }),
    compact(instrument || 'Stem'),
    stemType || 'IsolatedStem',
    `v${String(version ?? 1).padStart(2, '0')}`,
    `${Math.floor((sampleRate ?? 48_000) / 1000)}k`,
    `${bitDepth ?? 32}b`,
  ]

  return `${parts.join('_')}.wav`
}

export function buildPreviewNames({ stems, sessionMeta, sampleRate, bitDepth }) {
  return stems.map((stem) =>
    buildName({
      catalogId: sessionMeta.catalog_id,
      composer: sessionMeta.composer,
      title: sessionMeta.title,
      instrument: stem.instrument,
      stemType: stem.stem_type,
      version: sessionMeta.version,
      sampleRate,
      bitDepth,
    }),
  )
}
